### PR TITLE
Move cloud drive detection out of DrivesManager and improve sidebar population

### PR DESCRIPTION
**Pull request by AndyMDoyle**
DrivesManager was pretty complicated and had to deal with enumerating local and cloud drives, and rendering them both in the sidebar. Keeping this all in sync was leading to duplication and performance issues.

This PR introduces ```CloudDrivesManager``` which takes over responsibility of identifying cloud providers on a system, and rendering these in a section of the sidebar titled "Cloud Drives".

The original ```DrivesManager``` is now a little simpler, and has a rewritten method to add/remove drives in the sidebar.

I have identified that some of the sync and duplication issues appeared to be caused by timing issues likely introduced by various async calls being made from constructors (which obviously are not async methods). To improve this situation, ```CloudDrivesManager```, ```DrivesManager```, and ```SettingsViewModel``` have been tweaked to use a factory constructor pattern. Now instead of ```new```ing up those objects and having async methods called and forgotten about, a new instance can be obtained by calling ```await ClassName.CreateInstance();``` which allows async methods to be called when these objects are setup.

I have seen some areas; particularly stemming from ```SettingsViewModel``` where I feel classes have more responsibility than they need, and have complicated dependencies. For example, ```SettingsViewModel``` is responsible for loading all drives, cloud drives, and a lot of other system information - making it more than just settings. By being responsible for loading all drives, it has a hard dependency on ```DrivesManager```, which in one or more methods directly calls methods on views. ```DrivesManager``` also depends on ```DriveItem``` which also calls the UI to perform updates or run code on the UI thread.

I started going down a deep whole trying to refactor some of this to make changes easier to test with unit tests, but these dependencies make it almost impossible. So I worked my way back out and have done this with the minimal changes. I have some ideas that might help based on what I've done with some Xamarin apps, but I don't know how feasible they will be with a UWP app.

@gave92, please give this a try and see if it stops you duplicated drive issue.

### CODE REVIEW

File: Files/Filesystem/CloudDrivesManager.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                    {
+                        //No drives - remove the header
+                        MainPage.SideBarItems.Remove(drivesSection);
+                    }
+
+                    if (drivesSection == null)
+                    {
+                        drivesSection = new HeaderTextItem()
+                        {
+                            Text = "Cloud Drives"
```
Comment (yaichenbaum) [inline Files/Filesystem/CloudDrivesManager.cs:103]: Is it posible to add this string to the en-US resource file?

Comment (AndyMDoyle) [reply Files/Filesystem/CloudDrivesManager.cs:103]: Of course. Does it matter if a string does not have translations in all of the other resource files?

Comment (yaichenbaum) [reply Files/Filesystem/CloudDrivesManager.cs:103]: @AndyMDoyle I will take care of generating translations in the other files.

Comment (AndyMDoyle) [reply Files/Filesystem/CloudDrivesManager.cs]: Done

**Comment by yaichenbaum** [comment]
@AndyMDoyle Can you put the cloud section above the drive section?

**Comment by yaichenbaum** [comment]
@gave92 Once we are changing how this works, handling the drives should probably moved completly out of SettingsViewMode, what do you think?

**Comment by AndyMDoyle** [comment]
> @AndyMDoyle Can you put the cloud section above the drive section?

@yair100 Almost finished testing the SharePoint integration so will send over screenshot of it both ways around so you can get an idea of what it looks like with a few more cloud drives listed.

**Comment by yaichenbaum** [comment]
@AndyMDoyle That sounds great! The goal is to eventually make the groups collapsible so I wouldn't be too worried about how many items are in each section.

File: Files/ViewModels/SettingsViewModel.cs [hunk 2]
```diff
@@ -28,17 +29,31 @@ public class SettingsViewModel : ObservableObject
     {
         private readonly ApplicationDataContainer localSettings = ApplicationData.Current.LocalSettings;
 
-        public DrivesManager DrivesManager { get; }
+        public DrivesManager DrivesManager
```
Comment (gave92) [inline Files/ViewModels/SettingsViewModel.cs]: I agree @yair100 we should probably remove the drives from settings, they don't have much business here

File: Files/ViewModels/SettingsViewModel.cs [hunk 3]
```diff
@@ -53,6 +68,26 @@ public SettingsViewModel()
...
+        {
+            var settings = new SettingsViewModel(false);
+            return settings.Initialize();
+        }
+
+        private SettingsViewModel() : this(true)
+        {
+        }
+
+        private SettingsViewModel(bool initialize)
```
Comment (gave92) [inline Files/ViewModels/SettingsViewModel.cs:85]: At the end I'd keep only the constructor that's actually used

File: Files/Views/MainPage.xaml.cs [hunk 4]
```diff
@@ -75,7 +73,7 @@ protected override async void OnNavigatedTo(NavigationEventArgs eventArgs)
         {
             if (eventArgs.NavigationMode != NavigationMode.Back)
             {
-                App.AppSettings = new SettingsViewModel();
+                App.AppSettings = await SettingsViewModel.CreateInstance();
```
Comment (gave92) [inline Files/Views/MainPage.xaml.cs:76]: Mmm this adds 2 seconds of white screen before anything is loaded, not good.
If we remove drives from settings then loading drives should be done either at the end

File: Files/Views/YourHome.xaml.cs [hunk 5]
```diff
@@ -85,6 +85,18 @@ private async void RecentFilesWidget_RecentFileInvoked(object sender, UserContro
                             return;
                         }
                     }
+                    foreach (DriveItem drive in AppSettings.CloudDrivesManager.Drives)
```
Comment (gave92) [inline Files/Views/YourHome.xaml.cs:88]: Since the two for loops look identical we can just loop on the two lists together

**Review by gave92 — CHANGES_REQUESTED** [review]

**Comment by gave92** [comment]
Wait is it just me or the tab bar disappears with this PR? 😮

**Comment by AndyMDoyle** [comment]
Argh, I thought I had dealt with that tab bar issue by having settings available a little earlier.

It's again a timing issue because HorizontalMultitaskingControl relies on theme info from AppSettings. And while debugging to simulate a delay in loading settings, the MainPage view could possibly load before AppSettings (and the theme info) was populated, and if HorizontalMultitaskingControl can't get the settings it needs it doesn't appear properly.

I had a partial fix by moving the generation of AppSettings right to the startup of the app; but this caused issues because AppSettings depends on things that update the UI; and at that stage of the lifecycle the UI wasn't yet ready.

I will do some more on it this evening after work.

File: Files/Filesystem/CloudDrivesManager.cs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                    foreach (var drive in Drives.OrderBy(o => o.Text))
+                    {
+                        MainPage.SideBarItems.Insert(insertAt, drive);
+                        insertAt++;
+                    }
+                }
+            });
+        }
+
+        public static async Task<StorageFolderWithPath> GetRootFromPathAsync(string devicePath)
```
[the remark this answers is not in the document]
Comment (AndyMDoyle) [reply Files/Filesystem/CloudDrivesManager.cs:144]: Correct

File: Files/Filesystem/Drives.cs [hunk 7]
```diff
@@ -71,53 +85,91 @@ private async void EnumerateDrives()
...
+                            .Where(x => x.ItemType == NavigationControlItemType.Drive)
+                            .ToList())
                         {
                             MainPage.SideBarItems.Remove(item);
                             DrivesWidget.ItemsAdded.Remove(item);
                         }
+
+                        //Add all drives to the sidebar
+                        var insertAt = sectionStartIndex + 1;
+                        foreach (var drive in Drives.OrderBy(o => o.Text))
```
[the remark this answers is not in the document]
Comment (AndyMDoyle) [reply Files/Filesystem/Drives.cs]: Fixed

File: Files/ViewModels/SettingsViewModel.cs (hunk 2 above)
Comment (AndyMDoyle) [reply Files/ViewModels/SettingsViewModel.cs:32]: Maybe a challenge for another day :)

File: Files/ViewModels/SettingsViewModel.cs (hunk 3 above)
Comment (AndyMDoyle) [reply Files/ViewModels/SettingsViewModel.cs:85]: Removed

File: Files/Views/YourHome.xaml.cs (hunk 5 above)
Comment (AndyMDoyle) [reply Files/Views/YourHome.xaml.cs:88]: One is looping over DrivesManager, and the other CloudDrivesManager.

**Comment by AndyMDoyle** [comment]
@yair100 @gave92, please give this update a try to see if it addresses the tab bar issue on your system.

The population of AppSettings, InteractionViewModel, and SidebarPinnedController are now performed from App.xaml.cs when the ```OnLaunched``` and ```OnActivated``` events occur, and when the app is resumed from a suspended state. If these objects are already populated the work isn't done again to improve performance.

InteractionViewModel, SidebarPinnedController, and TerminalController have also been refactored slightly to allow async initialization to ensure these sort of dependencies are fully constructed before the application startup gets to the point where they are required.

Other changes in this update...

- Cloud Drives have been moved above Drives in thee sidebar
- Local drives retain the original order
- PR comments addressed

**Review by yair100 — DISMISSED** [review]

**Comment by yaichenbaum** [comment]
[CodeFactor](https://www.codefactor.io/repository/github/files-community/files) found an issue: A closing curly bracket must not be preceded by a blank line.

It's currently on:
[Files\Filesystem\CloudDrivesManager.cs:191
](https://github.com/files-community/files/blob/11c40efe64690c5cd71eb82b5915dfca81c2233a/Files/Filesystem/CloudDrivesManager.cs#L191)Commit 11c40efe64690c5cd71eb82b5915dfca81c2233a

**Comment by yaichenbaum** [comment]
[CodeFactor](https://www.codefactor.io/repository/github/files-community/files) found an issue: A closing curly bracket must not be preceded by a blank line.

It's currently on:
[Files\Filesystem\Drives.cs:160
](https://github.com/files-community/files/blob/11c40efe64690c5cd71eb82b5915dfca81c2233a/Files/Filesystem/Drives.cs#L160)Commit 11c40efe64690c5cd71eb82b5915dfca81c2233a

**Comment by AndyMDoyle** [comment]
> [CodeFactor](https://www.codefactor.io/repository/github/files-community/files) found an issue: A closing curly bracket must not be preceded by a blank line.
> 
> It's currently on:
> [Files\Filesystem\CloudDrivesManager.cs:191 ](https://github.com/files-community/files/blob/11c40efe64690c5cd71eb82b5915dfca81c2233a/Files/Filesystem/CloudDrivesManager.cs#L191)Commit [11c40ef](https://github.com/files-community/Files/commit/11c40efe64690c5cd71eb82b5915dfca81c2233a)

Resolved by removing an unused method earlier.

**Comment by AndyMDoyle** [comment]
> [CodeFactor](https://www.codefactor.io/repository/github/files-community/files) found an issue: A closing curly bracket must not be preceded by a blank line.
> 
> It's currently on:
> [Files\Filesystem\Drives.cs:160 ](https://github.com/files-community/files/blob/11c40efe64690c5cd71eb82b5915dfca81c2233a/Files/Filesystem/Drives.cs#L160)Commit [11c40ef](https://github.com/files-community/Files/commit/11c40efe64690c5cd71eb82b5915dfca81c2233a)

Fixed

File: Files/Views/MainPage.xaml.cs (hunk 4 above)
Comment (AndyMDoyle) [reply Files/Views/MainPage.xaml.cs]: @gave92 does the latest commit give you the same white screen?
Creatings SettingsViewModel is expensive, and I am already testing some ideas to split DrivesManager and CloudDrivesManager out and make them properties of App; with all dependencies being constructed in parallel to minimise startup time.

**Comment by AndyMDoyle** [comment]
@gave92 @yair100, is there anything else you need from me for this PR?

Once merged, I have a new PR to raise based on this that has improved startup performance a lot. On my system it's probably less than 2 seconds from clicking the icon to seeing everything in Release mode (debug mode is still around 8 seconds and most of that is down to XAML rendering). I have done this by moving DrivesManager and CloudDrivesManager out of SettingsViewModel which shaved off a little time, but the big improvement came from async instantiating CloudDriveManager and async loading cloud drive providers.

**Comment by gave92** [comment]
@AndyMDoyle looks good, no duplicates, no white screen, tabs are back.
Opened a small PR with some cleanup, let me know what you think.

**Comment by AndyMDoyle** [comment]
@gave92, your cleanup PR looks good. I have added one minor fix in addition as I found if the app suspends there was an issue with the drive watcher when resuming.

**Review by gave92 — APPROVED** [review]

**Comment by yaichenbaum** [comment]
@AndyMDoyle Thank you for your contribution 🚀

**Pull request merged**